packages/cozy-realtime/src/index.js

```diff
@@ -13,28 +13,59 @@ const WEBSOCKET_STATE = {
 const NUM_RETRIES = 3
 const RETRY_BASE_DELAY = 1000
 
+// stored subscriptions arguments to rerun while a retry
+// stored as a Map { [doctype]: socket }
+let subscriptionsState = new Set()
+
+export const getSubscriptionsState = () => subscriptionsState
+
+// listener key computing, according to doctype only or with doc id
+const LISTENER_KEY_SEPARATOR = '/' // safe since we can't have a '/' in a doctype
+const getListenerKey = (doctype, docId) =>
+  docId ? [doctype, docId].join(LISTENER_KEY_SEPARATOR) : doctype
+
+const getTypeAndIdFromListenerKey = listenerKey => {
+  const splitResult = listenerKey.split(LISTENER_KEY_SEPARATOR)
+  return {
+    doctype: splitResult[0],
+    docId: splitResult.length > 1 ? splitResult[1] : null
+  }
+}
+
 // Send a subscribe message for the given doctype trough the given websocket, but
 // only if it is in a ready state. If not, retry a few milliseconds later.
-function subscribeWhenReady(doctype, socket) {
+const MAX_SOCKET_POLLS = 500 // to avoid infinite poling
+export function subscribeWhenReady(
+  doctype,
+  socket,
+  docId,
+  remainedTries = MAX_SOCKET_POLLS
+) {
   if (socket.readyState === WEBSOCKET_STATE.OPEN) {
     try {
+      const payload = { type: doctype }
+      if (docId) payload.id = docId
       socket.send(
         JSON.stringify({
           method: 'SUBSCRIBE',
-          payload: {
-            type: doctype
-          }
+          payload
         })
       )
     } catch (error) {
-      // eslint-disable-next-line no-console
       console.warn(`Cannot subscribe to doctype ${doctype}: ${error.message}`)
       throw error
     }
   } else {
-    setTimeout(() => {
-      subscribeWhenReady(doctype, socket)
-    }, 10)
+    // no retries remaining
+    if (!remainedTries) {
+      const error = new Error('socket failed to connect')
+      console.warn(`Cannot subscribe to doctype ${doctype}: ${error.message}`)
+      throw error
+    } else {
+      setTimeout(() => {
+        subscribeWhenReady(doctype, socket, docId, --remainedTries)
+      }, 10)
```

Comment:
I don't like this polling. Can't we just to `socket.onOpen(() =>subscribeWhenReady(doctype, socket, docId))`?

Reply:
Or maybe we can use a promise resolved after sending the authentication, to avoid sending the subscribe message before the auth message.

Reply:
I added a promise to handle that properly, good point 👍